scalafix-nsc/src/main/scala/scalafix/nsc/NscSemanticApi.scala

```diff
@@ -11,104 +11,402 @@ import scalafix.ScalafixConfig
-        case _ =>
-          gtree.children.foldLeft(ctx)(evaluate)
+    }
+  }
+
+  private class OffsetTraverser extends g.Traverser {
+    val offsets = mutable.Map[Int, g.Tree]()
+    val treeOwners = mutable.Map[Int, g.Tree]()
+    override def traverse(t: g.Tree): Unit = {
+      t match {
+        case g.ValDef(_, name, tpt, _) if tpt.nonEmpty =>
+          offsets += (tpt.pos.point -> tpt)
+          treeOwners += (t.pos.point -> t)
+        case g.DefDef(_, _, _, _, tpt, _) =>
+          offsets += (tpt.pos.point -> tpt)
+          treeOwners += (t.pos.point -> t)
+        case _ => super.traverse(t)
       }
     }
-    evaluate(SemanticContext("", Nil), gtree)
-    builder
   }
 
   private def getSemanticApi(unit: g.CompilationUnit,
                              config: ScalafixConfig): SemanticApi = {
-    val offsets = offsetToType(unit.body, config.dialect)
+    // Compute scopes for global and local imports
+    val st = new ScopeTraverser
+    st.traverse(unit.body)
+    val rootPkg = st.topLevelPkg
+    val rootImported = st.scopes(rootPkg)
+    val realRootScope = rootPkg.info.members
+
+    // Compute offsets for the whole compilation unit
+    val traverser = new OffsetTraverser
+    traverser.traverse(unit.body)
+
+    def toMetaType(tp: g.Tree) =
+      config.dialect(tp.toString).parse[m.Type].get
+
+    def gimmePosition(t: m.Tree): m.Position = {
+      t match {
+        case m.Defn.Val(_, Seq(pat), _, _) => pat.pos
+        case m.Defn.Def(_, name, _, _, _, _) => name.pos
+        case _ => t.pos
+      }
+    }
+
+    /** Strip `this` references and convert indeterminate names to term names. */
+    def stripThis(ref: m.Tree) = {
+      ref.transform {
+        case m.Term.Select(m.Term.This(ind: m.Name.Indeterminate), name) =>
+          m.Term.Select(m.Term.Name(ind.value), name)
+        case m.Type.Select(m.Term.This(ind: m.Name.Indeterminate), name) =>
+          m.Type.Select(m.Term.Name(ind.value), name)
+      }
+    }
+
+    /** Look up a Meta name for both Reflect Term and Type names.
+      *
+      * Unfortunately, meta selection chains are term refs while they could
+      * be type refs (e.g. `A` in `A.this.B` is a term). For this reason,
+      * we need to check both names to guarantee whether a name is in scope.
+      */
+    @inline
+    def lookupBothNames(name: String,
+                        in: g.Scope,
+                        disambiguatingOwner: Option[g.Symbol],
+                        disambiguatingNamespace: String): g.Symbol = {
+      val typeName = g.TypeName(name)
+      val typeNameLookup = in.lookup(typeName)
+      val symbol =
+        if (typeNameLookup.exists) typeNameLookup
+        else {
+          val termName = g.TermName(name)
+          val termNameLookup = in.lookup(termName)
+          if (termNameLookup.exists) termNameLookup
+          else g.NoSymbol
+        }
+
+      // Disambiguate overloaded symbols caused by name shadowing
+      if (symbol.isOverloaded) {
```

Comment:
We try to disambiguate the overloaded symbol with the owner. If there's not luck, we fallback to a string-based approach: get the symbol whose full name is the longest common subexpression of our `tpe`.